mobify/platform-scaffold
# Loads scripts synchronously to improve load performance

Loads scripts synchronously to improve load performance

 **JIRA**: N/A
 **Linked PRs**: https://github.com/mobify/progressive-web-sdk/pull/673

## Changes
- Loads scripts synchronously in order for the browser to increase the priority of the scripts. The delay due to not being high priority can be up to 5 seconds on a Moto G4.
- Moves a number of things that the tag used to be responsible for into the loader:
  - Adds preview logic into loader. Previously, this was the job of the tag.
  - Adds the "opt-in" logic to the loader. This means going forward if we wanted to opt-in more browsers, we could do it without a tag change.
  - Adds a.js loading for non-supported browsers to the loader, since this is no longer something the tag is responsible for. 

## Notes
- The loader will not parse on IE8 or lower. However, the failure does not prevent the page from continuing to load, or other scripts from running. Tested on SauceLabs with IE6, 7 and 8.


## Loader blocking TODOs
- [x] Fix loading push
- [x] Make loader backwards compatible with V7 tag
- [x] Add conditional logic for loading asynchronously in order for the build to work on 2G (on Chrome, you cannot document.write when on a 2G connection)

## Loader non-blocking TODOs
- [ ] Refactor logic into SDK

## How to test-drive this PR
You'll need to use Charles to replace the existing tag on Merlin's with the new V8 tag.
- First, ensure you have SSL proxying enabled for www.merlinspotions.com (Proxy -> SSL Proxying Settings -> "Enable SSL Proxying", then add "www.merlinspotions.com").
- Add a rewrite rule (Tools -> Rewrite...), and match the following screenshot (notice how basic the tag is!):

<img width="1435" alt="2017-06-07 06 10 28 pm" src="https://user-images.githubusercontent.com/81582/26907997-a40b2770-4bac-11e7-99c6-19fb20994e29.png">

- You also need to map remote `https://cdn.mobify.com/sites/progressive-web-scaffold/production/*` to `https://localhost:8443/`, since the current live bundle won't have any of the preview logic we need to test. Match the following screenshot:

<img width="656" alt="2017-06-07 06 09 57 pm" src="https://user-images.githubusercontent.com/81582/26908054-025480b0-4bad-11e7-9328-396ff513b83d.png">

- Checkout [this branch](https://github.com/mobify/progressive-web-sdk/pull/673) of the SDK locally, run `npm link && npm run dev:build` in that folder, then in `platform-scaffold/web`, run `npm link progressive-web-sdk`.
- Then run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Everything should be working as normal! You should also see in the Network tab that requests should be loaded with "High" or "Medium" priority (to do this, enable the priority column in the Network tab).
- Then you'll need to test with a few other combinations:
  - Try without preview. The site should still work, and priority should still be High/Medium.
  - Try with preview, but with the "Rewrite" rule disabled, to ensure this build still works with the old tag (priority will be low, since it's the old synchronous style of loading).
  - Try without preview, still with "Rewrite" rule disabled, to ensure this build still works with the old tag when not previewing (priority will be low, since it's the old synchronous style of loading).
